prusti-specs/src/type_model/mod.rs

```diff
@@ -0,0 +1,392 @@
+    item_struct: &syn::ItemStruct,
+    generated_model_struct: &syn::ItemStruct,
+    to_model_trait: &syn::ItemTrait,
+) -> TypeModelGenerationResult<syn::ItemImpl> {
+    let ident = &item_struct.ident;
+    let model_struct_ident = &generated_model_struct.ident;
+    let mut rewritten_generics: Vec<syn::GenericParam> = Vec::new();
+    for param in &item_struct.generics.params {
+        match param {
+            syn::GenericParam::Lifetime(_) => rewritten_generics.push(parse_quote!('_)),
+            syn::GenericParam::Type(type_param) => {
+                rewritten_generics.push(syn::GenericParam::Type(type_param.clone()))
+            }
+            syn::GenericParam::Const(const_param) => {
+                return Err(TypeModelGenerationError::ConstParamDisallowed(
+                    const_param.span(),
+                ))
+            }
+        }
+    }
+
+    let impl_path: syn::Path = parse_quote!(
+        #ident < #(#rewritten_generics),* >
+    );
+
+    // Create impl for provided trait
+    let to_model_trait_ident = &to_model_trait.ident;
+    Ok(parse_quote_spanned! {item_struct.span()=>
+        impl #to_model_trait_ident for #impl_path {
+            #[trusted]
+            #[pure]
+            #[prusti::model_generator]
+            fn model(&self) -> #model_struct_ident {
+                unimplemented!("Models can only be used in specifications")
+            }
+        }
+    })
+}
+
+/// [syn::Ident]s which are used for the generated items
+struct GeneratedIdents {
+    model_struct_ident: syn::Ident,
+    to_model_trait_ident: syn::Ident,
+}
+
+impl GeneratedIdents {
+    fn generate(item_struct: &syn::ItemStruct) -> Self {
+        let mut name = item_struct.ident.to_string();
+
+        for param in item_struct.generics.params.iter() {
+            if let syn::GenericParam::Type(ty_param) = param {
+                name.push_str(ty_param.ident.to_string().as_str());
+            }
+        }
+
+        GeneratedIdents {
+            model_struct_ident: Ident::new(
+                format!("Prusti{}Model", name).as_str(),
+                item_struct.ident.span(),
+            ),
+            to_model_trait_ident: Ident::new(
+                format!("Prusti{}ToModel", name).as_str(),
+                item_struct.ident.span(),
+            ),
+        }
+    }
+}
+
+/// Errors that can happen during model generation.
+/// These are mostly wrong usages of the macro by the user.
+#[derive(Debug)]
+enum TypeModelGenerationError {
+    /// Thrown when the model contains no fields
+    MissingStructFields(proc_macro2::Span),
+
+    /// Thrown when using a const type param in the model
+    ConstParamDisallowed(proc_macro2::Span),
+}
+
+impl std::convert::From<TypeModelGenerationError> for syn::Error {
+    fn from(err: TypeModelGenerationError) -> Self {
+        match err {
+            TypeModelGenerationError::MissingStructFields(span) => {
+                syn::Error::new(span, "Missing fields in model")
```

Comment:
This error message is not great: it could be misunderstood as saying there are some *particular* fields which should be in the model. How about "Type model must have at least one field"?

Reply:
Your suggestion is good :+1: 

If I recall correctly, there is no technical reason for this error. Imo, it just makes no sense to have a model without fields.